keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+
+	// Information for when was the next time the job will be scheduled.
+    // May not be accurate in some cases for Forbid concurrency policy
+	// +optional
+	NextScheduleTime *metav1.Time
+
+	// Information when was the last time the job successfully completed.
+	// +optional
+	LastSuccessfulTime *metav1.Time
+}
+```
+
+### CronJob v1beta1 API
+
+All the new fields described in the v1 section would be introduced in the v1beta1 API as well.
+
+### Validations
+
+Nothing additional from v1beta1
+
+## Tests
+
+### E2E test
+
+CronJob E2E test code is [located here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/cronjob.go). The new controller should pass the current set of E2E.
+
+#### Existing test cases
+
+- ConcurrencyPolicy
+  - should schedule multiple jobs concurrently
+  - should not schedule new jobs when ForbidConcurrent
+  - should replace jobs when ReplaceConcurrent
+- Suspend
+  - should not schedule jobs when suspended
+- SuccessfulJobsHistoryLimit
+  - should delete successful finished jobs when above successfulJobsHistoryLimit
+- FailedJobsHistoryLimit
+  - should delete failed finished jobs when above failedJobsHistoryLimit
+- Events Recorder
+  - should not emit unexpected warnings
+  - should remove from active list jobs that have been deleted
+
+#### New test cases
+
+- Schedule
+  - Should not create a cronjob with invalid schedule format
+- StartingDeadlineSeconds
+  - Should not schedule a job within two minutes when missed the current window if StartingDeadlineSeconds is 0
+  - Should schedule a job soon when missed the current window if StartingDeadlineSeconds is long
+- JobTemplate
+  - Should not schedule a job with invalid job template
+- [Endpoints coverage](https://apisnoop.cncf.io/?zoomed=category-beta-batch)
+  - Should list cronjobs for all namespaces
+  - Should update a cronjob
+  - Should patch a cronjob
+  - Should delete all cronjobs in a namespace
+  - Should get a cronjob status
+  - Should update a cronjob status
+  - Should patch a cronjob status
+- Tests covering Bug fixes
+  - [issue/63371 - Should not start “missed” jobs from old cronjob after updating time](https://github.com/kubernetes/kubernetes/issues/63371)
+  - [issue/74741 - Should cleanup finished pods when job is completed](https://github.com/kubernetes/kubernetes/issues/74741)
+  - [issue/76570 - Should not keep creating new pods when job image has ImagePullBackOff error](https://github.com/kubernetes/kubernetes/issues/76570)
+  - [issue/78646 - Should change schedule when schedule value is not wrapped with quotes](https://github.com/kubernetes/kubernetes/issues/78646)
+- Scaling
+  - Should be able to create and schedule at least 5000 cronjobs
+  - Measure scheduling skew, podCreationLatency and check if it meets expectation (To Be Defined)
+- Start Stop Tests
+  - Schedule cronjobs and randomly stop the controller and start it.
+  - Schedule cronjobs and stop the controller and start it after the deadline.
+
```

Comment:
Should there be a test for having cronjobs switching between 2 controllers (toggling the feature flag)?

Reply:
Effectively there will never be two different controllers running at the same time (even in HA setup only one will be the leader).
So how would this test be different from actual "controller restart"?

Reply:
We need to ensure to run full e2e with both, I need to sync with sig-testing how to set it up, but never we'll be running these two at the same time.